src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,445 @@
+@format_option
+@sdk_options()
+def list_devices(
+    state,
+    active,
+    days_since_last_connected,
+    drop_most_recent,
+    org_uid,
+    include_backup_usage,
+    include_usernames,
+    include_settings,
+    format,
+):
+    """Outputs a list of all devices."""
+    columns = [
+        "computerId",
+        "guid",
+        "name",
+        "osHostname",
+        "status",
+        "lastConnected",
+        "productVersion",
+        "osName",
+        "osVersion",
+        "userUid",
+    ]
+    devices_dataframe = _get_device_dataframe(
+        state.sdk, columns, active, org_uid, include_backup_usage
+    )
+    if drop_most_recent:
+        devices_dataframe = _drop_n_devices_per_user(
+            devices_dataframe, drop_most_recent
+        )
+    if days_since_last_connected:
+        devices_dataframe = _drop_devices_which_have_not_connected_in_some_number_of_days(
+            devices_dataframe, days_since_last_connected
+        )
+    if include_settings:
+        devices_dataframe = _add_settings_to_dataframe(state.sdk, devices_dataframe)
+    if include_usernames:
+        devices_dataframe = _add_usernames_to_device_dataframe(
+            state.sdk, devices_dataframe
+        )
+    formatter = DataFrameOutputFormatter(format)
+    formatter.echo_formatted_dataframe(devices_dataframe)
+
+
+def _get_device_dataframe(
+    sdk, columns, active=None, org_uid=None, include_backup_usage=False
+):
+    devices_generator = sdk.devices.get_all(
+        active=active, include_backup_usage=include_backup_usage, org_uid=org_uid
+    )
+    devices_list = []
+    if include_backup_usage:
+        columns.append("backupUsage")
+    for page in devices_generator:
+        devices_list.extend(page["computers"])
+    return DataFrame.from_records(devices_list, columns=columns)
+
+
+def _add_settings_to_dataframe(sdk, device_dataframe):
+    macos_guids = device_dataframe.loc[
+        device_dataframe["osName"] == "mac", "guid"
+    ].values
+
+    def handle_row(guid):
+        try:
+            full_disk_access_status = sdk.devices.get_agent_full_disk_access_state(
+                guid
+            ).data[
+                "value"
+            ]  # returns 404 error if device isn't a Mac or doesn't have full disk access
+        except Py42NotFoundError:
+            full_disk_access_status = False
+        return {
+            "guid": guid,
+            "full disk access status": full_disk_access_status,
+        }
+
+    result_list = DataFrame.from_records(
+        run_bulk_process(
+            handle_row, macos_guids, progress_label="Getting device settings"
+        )
+    )
+    try:
+        return device_dataframe.merge(result_list, how="left", on="guid")
+    except KeyError:
+        return device_dataframe
+
+
+def _drop_devices_which_have_not_connected_in_some_number_of_days(
+    devices_dataframe, days_since_last_connected
+):
+    return devices_dataframe.loc[
+        to_datetime(datetime.now(), utc=True)
```

Comment:
This doesn't actually convert the time to UTC time, it just sets the timezone info to UTC. 

```
>>> from datetime import datetime
>>> from pandas import to_datetime
>>> 
>>> now = datetime.now()
>>> now
datetime.datetime(2020, 12, 29, 14, 48, 43, 395819)
>>> to_datetime(now, utc=True)
Timestamp('2020-12-29 14:48:43.395819+0000', tz='UTC')
```

We'll want to use `datetime.utcnow()`. 

Also, can we turn these datetime conversions into variables to make the `.loc[]` logic more readable?

```
    now = to_datetime(datetime.utcnow(), utc=True)
    last_connected = to_datetime(devices_dataframe["lastConnected"], utc=True)
    delta = to_timedelta(days_since_last_connected, unit="days")
    return devices_dataframe.loc[now - last_connected > delta]
```

Reply:
done! I removed the `utc=True` from the last_connected line as well, since to_timedelta does retrieve the time zone info from the timestamps our console provides. This way we will correctly raise an error if console starts giving us timestamps in non-UTC timezones.

Reply:
We need to keep that on the last_connected conversion, otherwise the subtraction won't work:

```
>>> last_connected = '2019-12-16T17:31:31.161-06:00'
>>> to_datetime(datetime.utcnow(), utc=True) - to_datetime(last_connected)
Traceback (most recent call last):
  File "<input>", line 1, in <module>
    to_datetime(datetime.utcnow(), utc=True) - to_datetime(last_connected)
  File "pandas/_libs/tslibs/timestamps.pyx", line 364, in pandas._libs.tslibs.timestamps._Timestamp.__sub__
TypeError: Timestamp subtraction must have the same timezones or no timezones
```

Reply:
Oh, I see, it does do the conversion there. Ok, I've added it back.